### PR TITLE
【Hackathon4 No58】logcumsum logsum

### PR DESCRIPTION
**Pull request by enkilee**
### PR types

Bug fixes
### PR changes

APIs
### Describe

fix logcumsum: add bfloat support and test
fix logsum: add bfloat support and test

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_logcumsumexp_op.py [hunk 1]
```diff
@@ -298,13 +298,40 @@ def test_main(self):
...
 
         y_np_1, x_g_np_1 = self.check_main(x_np, 'float16', axis=1)
         y_np_2, x_g_np_2 = self.check_main(x_np, 'float32', axis=1)
-        np.testing.assert_allclose(y_np_1, y_np_2, rtol=1e-03)
-        np.testing.assert_allclose(x_g_np_1, x_g_np_2, rtol=2e-03)
+        np.testing.assert_allclose(y_np_1, y_np_2, rtol=0.5)
+        np.testing.assert_allclose(x_g_np_1, x_g_np_2, rtol=0.5)
+
+
+class TestLogcumsumexpBfloat16(unittest.TestCase):
```
Comment (Vvsmile) [inline python/paddle/fluid/tests/unittests/test_logcumsumexp_op.py]: 改一下类名，直接让class继承OpTest就好，下面的def ……BF16（OpTest）可以去掉

File: python/paddle/fluid/tests/unittests/test_logsumexp.py [hunk 2]
```diff
@@ -182,7 +182,22 @@ def test_check_grad(self):
         ref_x_grad = logsumexp_ref_grad(ref_x)
         x = self.inputs['X'].astype(np.float16)
         x_grad = logsumexp_op_grad(x)
-        np.testing.assert_allclose(x_grad, ref_x_grad, rtol=1e-03, atol=1e-05)
+        np.testing.assert_allclose(x_grad, ref_x_grad, rtol=1e-02, atol=1e-02)
+
+
+class TestLogsumexpOpBP16(unittest.TestCase):
```
Comment (Vvsmile) [inline python/paddle/fluid/tests/unittests/test_logsumexp.py]: 直接继承OpTest框架

File: python/paddle/fluid/tests/unittests/test_logsumexp.py [hunk 3]
```diff
@@ -182,7 +182,22 @@ def test_check_grad(self):
...
+    def TestLogcumsumexpbp16(OpTest):
+        def setUp(self):
+            self.op_type = 'logsumexp'
+            self.dtype = np.uint16
+            x = np.random.rand([10, 12]).astype(np.float32)
+            out = ref_logsumexp(x)
+            self.inputs = {'X': convert_float_to_uint16(x)}
+            tensor_x = paddle.to_tensor(convert_float_to_uint16(x))
+            out_pad = logsumexp_wrapper(tensor_x)
+            np.testing.assert_allclose(
```
Comment (Vvsmile) [inline python/paddle/fluid/tests/unittests/test_logsumexp.py]: 继承之后改成check_output 及 check_grad调用，内部会自动进行测试，前面继承OpTest的同理

File: python/paddle/fluid/tests/unittests/test_logsumexp.py [hunk 4]
```diff
@@ -182,7 +182,32 @@ def test_check_grad(self):
...
-        np.testing.assert_allclose(x_grad, ref_x_grad, rtol=1e-03, atol=1e-05)
+        np.testing.assert_allclose(x_grad, ref_x_grad, rtol=1e-02, atol=1e-02)
+
+
+@unittest.skipIf(
+    not core.is_compiled_with_cuda()
+    or not core.is_bfloat16_supported(core.CUDAPlace(0)),
+    "core is not complied with CUDA and not support the bfloat16",
+)
+class TestLogsumexpBF16(OpTest):
```
Comment (Vvsmile) [inline python/paddle/fluid/tests/unittests/test_logsumexp.py]: 类名改成TestLogsumexpBF16Op吧，之前规范里给的有问题麻烦了

File: python/paddle/fluid/tests/unittests/test_logcumsumexp_op.py [hunk 5]
```diff
@@ -298,13 +298,38 @@ def test_main(self):
...
+        np.testing.assert_allclose(y_np_1, y_np_2, rtol=0.5)
+        np.testing.assert_allclose(x_g_np_1, x_g_np_2, rtol=0.5)
+
+
+@unittest.skipIf(
+    not core.is_compiled_with_cuda()
+    or not core.is_bfloat16_supported(core.CUDAPlace(0)),
+    "core is not complied with CUDA and not support the bfloat16",
+)
+class TestLogcumsumexpBF16(OpTest):
```
Comment (Vvsmile) [inline python/paddle/fluid/tests/unittests/test_logcumsumexp_op.py]: 类名改成TestLogcumsumexpBF16Op吧

**Comment by luotao1** [comment]
请解决下windows-inference流水线上`test_logcumsumexp_op`失败的问题

File: python/paddle/fluid/tests/unittests/test_logcumsumexp_op.py [hunk 6]
```diff
@@ -314,5 +314,46 @@ def test_main(self):
...
+
+    def test_check_output(self):
+        self.check_output()
+
+    def test_check_grad(self):
+        self.inputs = {'X': input}
+        self.check_grad(
+            ['X'],
+            'Out',
+            user_defined_grads=[
```
Comment (Vvsmile) [inline python/paddle/fluid/tests/unittests/test_logcumsumexp_op.py]: 可以先尝试一下不用user_defined_grads

Comment (enkilee) [reply python/paddle/fluid/tests/unittests/test_logcumsumexp_op.py]: AssertionError: 
2023-04-04 16:57:01 Not equal to tolerance rtol=0.01, atol=0.1
2023-04-04 16:57:01 Operator logcumsumexp error, Gradient Check On Place(gpu:0) variable X (shape: (10, 10), dtype: <class 'numpy.uint16'>) max gradient diff over limit
2023-04-04 16:57:01 Mismatched elements: 2 / 100 (2%)
2023-04-04 16:57:01 Max absolute difference: 49.390625
2023-04-04 16:57:01 Max relative difference: 0.9878125
2023-04-04 16:57:01  x: array([[0.000000e+00, 0.000000e+00, 0.000000e+00, 0.000000e+00,
2023-04-04 16:57:01         0.000000e+00, 0.000000e+00, 0.000000e+00, 0.000000e+00,
2023-04-04 16:57:01         0.000000e+00, 9.183550e-41],...
2023-04-04 16:57:01  y: array([[ 0.      ,  0.      ,  0.      ,  0.      ,  0.      ,  0.      ,
2023-04-04 16:57:01          0.      ,  0.      ,  0.      ,  0.      ],
2023-04-04 16:57:01        [ 0.      ,  0.      ,  0.      ,  0.      ,  0.      ,  0.      ,...

Comment (enkilee) [reply python/paddle/fluid/tests/unittests/test_logcumsumexp_op.py]: 请问这种，该如何计算误差阈值

**Comment by enkilee** [comment]
self.check_output_with_place(place, atol=1e-3):
ValueError: operands could not be broadcast together with shapes (10,10) (100,)  请问这种应该如何修改

File: python/paddle/tensor/math.py [hunk 7]
```diff
@@ -3303,7 +3303,7 @@ def logcumsumexp(x, axis=None, dtype=None, name=None):
     Args:
         x (Tensor): The input tensor.
         axis (int, optional): The dimension to do the operation along. -1 means the last dimension. The default (None) is to compute the cumsum over the flattened array.
-        dtype (str, optional): The data type of the output tensor, can be float16, float32, float64. If specified, the input tensor is casted to dtype before the operation is performed. This is useful for preventing data type overflows. The default value is None.
+        dtype (str, optional): The data type of the output tensor, can be float16, float32, float64, uint16. If specified, the input tensor is casted to dtype before the operation is performed. This is useful for preventing data type overflows. The default value is None.
```
Comment (ZzSean) [inline python/paddle/tensor/math.py]: 文档暂时不用修改

**Review by ZzSean — DISMISSED** [review]
LGTM

File: python/paddle/tensor/math.py [hunk 8]
```diff
@@ -2163,7 +2163,7 @@ def logsumexp(x, axis=None, keepdim=False, name=None):
        logsumexp(x) = \log\sum exp(x)
 
     Args:
-        x (Tensor): The input Tensor with data type float16, float32 or float64, which
+        x (Tensor): The input Tensor with data type float16, float32, float64 or uint16, which
```
Comment (ZzSean) [inline python/paddle/tensor/math.py]: 这里文档还是有改动了。。。

**Review by ZzSean — APPROVED** [review]
LGTM

**Pull request merged**